File: shared/ui/ipc/host.protocol.editor.ts
import { RequestType } from "vscode-jsonrpc";
import { NotificationType } from "vscode-languageserver-protocol";
import { Position, Range } from "vscode-languageserver-types";
import { EditorContext, EditorSelection, IpcRoutes } from "./webview.protocol.common";

export interface GetActiveEditorContextResponse {
	editorContext: EditorContext;
}

export const GetActiveEditorContextRequestType = new RequestType<
	void,
	GetActiveEditorContextResponse,
	void,
	void
>(`${IpcRoutes.Host}/editor/context`);

export interface EditorHighlightRangeRequest {
	uri: string;
	range: Range;
	highlight: boolean;
}
export interface EditorHighlightRangeResponse {
	success: boolean;
}

export const EditorHighlightRangeRequestType = new RequestType<
	EditorHighlightRangeRequest,
	EditorHighlightRangeResponse,
	void,
	void
>(`${IpcRoutes.Host}/editor/range/highlight`);

export interface EditorRevealRangeRequest {
	uri: string;
	range: Range;
	preserveFocus?: boolean;
	atTop?: boolean;
}
export interface EditorRevealRangeResponse {
	success: boolean;
}
export const EditorRevealRangeRequestType = new RequestType<
	EditorRevealRangeRequest,
	EditorRevealRangeResponse,
	void,
	void
>(`${IpcRoutes.Host}/editor/range/reveal`);

export interface EditorGetRangeSymbolsRequest {
	uri: string;
	range: Range;
}

export interface EditorGetRangeSymbolsResponse {
	symbols: string[];
}

export const EditorGetRangeSymbolsRequestType = new RequestType<
	EditorGetRangeSymbolsRequest,
	EditorGetRangeSymbolsResponse,
	void,
	void
>(`${IpcRoutes.Host}/editor/range/symbols`);

export interface EditorSelectRangeRequest {
	uri: string;
	selection: EditorSelection;
	preserveFocus?: boolean;
}
export interface EditorSelectRangeResponse {
	success: boolean;
}
export const EditorSelectRangeRequestType = new RequestType<
	EditorSelectRangeRequest,
	EditorSelectRangeResponse,
	void,
	void
>(`${IpcRoutes.Host}/editor/range/select`);

export interface EditorScrollToNotification {
	uri: string;
	position: Position;
	deltaPixels?: number;
	atTop?: boolean;
}
export const EditorScrollToNotificationType = new NotificationType<
	EditorScrollToNotification,
	void
>(`${IpcRoutes.Host}/editor/scrollTo`);

File: vs/src/CodeStream.VisualStudio/Models/Host.Protocol.Editor.cs
using Microsoft.VisualStudio.LanguageServer.Protocol;
using System;

namespace CodeStream.VisualStudio.Models {
	public class EditorHighlightRangeResponse {
		public bool Success { get; set; }
	}

	public class EditorHighlightRangeRequest {
		public EditorHighlightRangeRequest(Uri uri, Range range, bool highlight) {
			Uri = uri.ToString();
			Range = range;
			Highlight = highlight;
		}

		public string Uri { get; }
		/// <summary>
		/// A single-line range with start & end char of 0 indicates a full-line highlight
		/// </summary>
		public Range Range { get; }
		public bool Highlight { get; }		
	}

	public class EditorHighlightRangeRequestType : RequestType<EditorHighlightRangeRequest> {
		public const string MethodName = "host/editor/range/highlight";
		public override string Method => MethodName;
	}

	public class EditorRevealRangeRequest {
		public EditorRevealRangeRequest(Uri uri, Range range, bool? preserveFocus, bool? atTop) {
			Uri = uri.ToString();
			Range = range;
			PreserveFocus = preserveFocus;
			AtTop = atTop;
		}

		public string Uri { get; }
		public Range Range { get; }
		/// <summary>
		/// True, if the focus of the webview should remain
		/// </summary>
		public bool? PreserveFocus { get; }
		public bool? AtTop { get; }
	}

	public class EditorRevealRangeResponse {
		public bool Success { get; set; }
	}

	public class EditorRevealRangeRequestType : RequestType<EditorRevealRangeRequest> {
		public const string MethodName = "host/editor/range/reveal";
		public override string Method => MethodName;
	}

	public class EditorSelectRangeResponse {
		public bool Success { get; set; }
	}

	public class EditorSelectRangeRequest {
		public EditorSelectRangeRequest(Uri uri, EditorSelection selection, bool? preserveFocus) {
			Uri = uri.ToString();
			Selection = selection;
			PreserveFocus = preserveFocus;
		}

		public string Uri { get; }
		/// <summary>
		/// A single-line range with start & end char of 0 indicates a full-line highlight
		/// </summary>
		public EditorSelection Selection { get; }
		/// <summary>
		/// True, if the focus of the webview should remain
		/// </summary>
		public bool? PreserveFocus { get; }
	}

	public class EditorSelectRangeRequestType : RequestType<EditorSelectRangeRequest> {
		public const string MethodName = "host/editor/range/select";
		public override string Method => MethodName;
	}
}
